In [2]:
import pandas as pd
import numpy as np

In [3]:
df_police = pd.read_csv('police.csv')
df_police.head()

stop_date stop_time  country_name driver_gender  driver_age_raw  \
0   1/2/2005      1:55           NaN             M          1985.0   
1  1/18/2005      8:15           NaN             M          1965.0   
2  1/23/2005     23:15           NaN             M          1972.0   
3  2/20/2005     17:15           NaN             M          1986.0   
4  3/14/2005     10:00           NaN             F          1984.0   

   driver_age driver_race     violation_raw violation  search_conducted  \
0        20.0       White          Speeding  Speeding             False   
1        40.0       White          Speeding  Speeding             False   
2        33.0       White          Speeding  Speeding             False   
3        19.0       White  Call for Service     Other             False   
4        21.0       White          Speeding  Speeding             False   

  search_type   stop_outcome is_arrested stop_duration  drugs_related_stop  
0         NaN       Citation       False      0-15 Min               False  
1         NaN       Citation       False      0-15 Min               False  
2         NaN       Citation       False      0-15 Min               False  
3         NaN  Arrest Driver        True     16-30 Min               False  
4         NaN       Citation       False      0-15 Min               False

In [4]:
df_police.shape

(65535, 15)

In [5]:
df_police.columns

Index(['stop_date', 'stop_time', 'country_name', 'driver_gender',
       'driver_age_raw', 'driver_age', 'driver_race', 'violation_raw',
       'violation', 'search_conducted', 'search_type', 'stop_outcome',
       'is_arrested', 'stop_duration', 'drugs_related_stop'],
      dtype='object')

In [7]:
df_police.dtypes

stop_date              object
stop_time              object
country_name          float64
driver_gender          object
driver_age_raw        float64
driver_age            float64
driver_race            object
violation_raw          object
violation              object
search_conducted         bool
search_type            object
stop_outcome           object
is_arrested            object
stop_duration          object
drugs_related_stop       bool
dtype: object

In [8]:
df_police.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 65535 entries, 0 to 65534
Data columns (total 15 columns):
 #   Column              Non-Null Count  Dtype  
---  ------              --------------  -----  
 0   stop_date           65535 non-null  object 
 1   stop_time           65535 non-null  object 
 2   country_name        0 non-null      float64
 3   driver_gender       61474 non-null  object 
 4   driver_age_raw      61481 non-null  float64
 5   driver_age          61228 non-null  float64
 6   driver_race         61475 non-null  object 
 7   violation_raw       61475 non-null  object 
 8   violation           61475 non-null  object 
 9   search_conducted    65535 non-null  bool   
 10  search_type         2479 non-null   object 
 11  stop_outcome        61475 non-null  object 
 12  is_arrested         61475 non-null  object 
 13  stop_duration       61475 non-null  object 
 14  drugs_related_stop  65535 non-null  bool   
dtypes: bool(2), float64(3), object(10)
memory usage: 6.6+

1. Remove the column that only contains missing values.

In [9]:
df_police.isnull().sum()

stop_date                 0
stop_time                 0
country_name          65535
driver_gender          4061
driver_age_raw         4054
driver_age             4307
driver_race            4060
violation_raw          4060
violation              4060
search_conducted          0
search_type           63056
stop_outcome           4060
is_arrested            4060
stop_duration          4060
drugs_related_stop        0
dtype: int64

Here, as seen above, country_name column has all values as null. In the sense, the entire column is null.
We can say all values because the number of rows is 65535 which is also the number of null values in country_name

In [10]:
df_police.drop(columns='country_name', axis=1, inplace=True)

In [13]:
df_police.columns.isin(['country_name'])

array([False, False, False, False, False, False, False, False, False,
       False, False, False, False, False])

2.  For Speeding , were Men or Women stopped more often ? 

In [14]:
df_police['violation'].unique()

array(['Speeding', 'Other', 'Equipment', 'Moving violation', nan,
       'Registration/plates', 'Seat belt'], dtype=object)

In [16]:
df_police['driver_gender'].unique()

array(['M', 'F', nan], dtype=object)

In [20]:
speeding = (df_police[df_police['violation'] == 'Speeding'])

In [26]:
no_of_men_speeding = len(speeding[speeding['driver_gender'] == 'M'])
no_of_men_speeding

25517

In [27]:
no_of_women_speeding = len(speeding[speeding['driver_gender'] == 'F'])
no_of_women_speeding

11686

So, we can say that there were more men who were speeding than women.


OR: another way

In [28]:
speeding['driver_gender'].value_counts()

driver_gender
M    25517
F    11686
Name: count, dtype: int64

3. Does gender affect who gets searched during a stop ?

In [35]:
df_police.columns

Index(['stop_date', 'stop_time', 'driver_gender', 'driver_age_raw',
       'driver_age', 'driver_race', 'violation_raw', 'violation',
       'search_conducted', 'search_type', 'stop_outcome', 'is_arrested',
       'stop_duration', 'drugs_related_stop'],
      dtype='object')

In [37]:
df_police.groupby('driver_gender')['search_conducted'].sum()

driver_gender
F     366
M    2113
Name: search_conducted, dtype: int64

In [39]:
pd.crosstab(df_police['driver_gender'], df_police['search_conducted'])

search_conducted  False  True 
driver_gender                 
F                 15944    366
M                 43051   2113

4. What is the mean stop_duration ?

In [41]:
df_police['stop_duration'].value_counts()

stop_duration
0-15 Min     47379
16-30 Min    11448
30+ Min       2647
2                1
Name: count, dtype: int64

In [52]:
df_police['stop_duration'] = df_police['stop_duration'].replace({
    '0-15 Min': 7.5,
    '16-30 Min': 24,
    '30+ Min': 45
})

In [53]:
df_police['stop_duration'].value_counts()

stop_duration
7.5     47379
24.0    11448
45.0     2647
Name: count, dtype: int64

In [55]:
df_police['stop_duration'].mean()

12.187420698181345

5. Compare the age distributions for each violation.

In [56]:
df_police['driver_age'].value_counts()

driver_age
22.0    2912
21.0    2886
20.0    2783
23.0    2725
24.0    2583
        ... 
15.0       5
84.0       3
88.0       2
83.0       2
85.0       1
Name: count, Length: 73, dtype: int64

In [57]:
df_police['violation'].value_counts()

violation
Speeding               37204
Moving violation       11926
Equipment               6516
Other                   3583
Registration/plates     2243
Seat belt                  3
Name: count, dtype: int64

In [65]:
df_age_violation = (df_police.groupby(by='violation')['driver_age'])
df_age_violation.describe()

count       mean        std   min   25%   50%   75%  \
violation                                                                    
Equipment             6507.0  31.682957  11.380671  16.0  23.0  28.0  39.0   
Moving violation     11876.0  36.736443  13.258350  15.0  25.0  35.0  47.0   
Other                 3477.0  40.362381  12.754423  16.0  30.0  41.0  50.0   
Registration/plates   2240.0  32.656696  11.150780  16.0  24.0  30.0  40.0   
Seat belt                3.0  30.333333  10.214369  23.0  24.5  26.0  34.0   
Speeding             37120.0  33.262581  12.615781  15.0  23.0  30.0  42.0   

                      max  
violation                  
Equipment            81.0  
Moving violation     86.0  
Other                86.0  
Registration/plates  74.0  
Seat belt            42.0  
Speeding             88.0